File: 8vim/src/main/java/inc/flide/vim8/keyboardactionlisteners/KeypadActionListener.java
package inc.flide.vim8.keyboardactionlisteners;

import static inc.flide.vim8.models.AppPrefsKt.appPreferenceModel;

import android.content.Context;
import android.media.AudioManager;
import android.os.Build;
import android.view.HapticFeedbackConstants;
import android.view.KeyEvent;
import android.view.View;
import inc.flide.vim8.MainInputMethodService;
import inc.flide.vim8.models.AppPrefs;
import inc.flide.vim8.models.CustomKeycode;
import inc.flide.vim8.models.KeyboardAction;
import inc.flide.vim8.models.LayerLevel;

public abstract class KeypadActionListener {
    public static final int KEYCODE_PROFILE_SWITCH = 288;
    private final AudioManager audioManager;
    protected MainInputMethodService mainInputMethodService;
    protected View view;

    public KeypadActionListener(MainInputMethodService mainInputMethodService, View view) {
        this.mainInputMethodService = mainInputMethodService;
        this.view = view;
        this.audioManager = (AudioManager) view.getContext().getSystemService(Context.AUDIO_SERVICE);
    }

    private boolean keyCodeIsValid(int keyCode) {
        int keycodeProfileSwitch = (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) ? KeyEvent.KEYCODE_PROFILE_SWITCH :
                KEYCODE_PROFILE_SWITCH;
        return keyCode >= KeyEvent.KEYCODE_UNKNOWN && keyCode <= keycodeProfileSwitch;
    }


    public void handleInputKey(int keyCode, int keyFlags) {
        boolean actionHandled = handleKeyEventKeyCodes(keyCode, keyFlags);
        if (!actionHandled) {
            CustomKeycode customKeycode = CustomKeycode.fromIntValue(keyCode);
            if (customKeycode != null) {
                actionHandled = customKeycode.handleKeyCode(mainInputMethodService);
            }
        }
        if (!actionHandled) {
            onText(String.valueOf((char) keyCode));
        } else {
            performInputAcceptedFeedback(keySound(keyCode));
        }
    }

    public float getCtrlAlpha() {
        return mainInputMethodService.getCtrlAlpha();
    }

    public void performCtrlToggle() {
        mainInputMethodService.performCtrlToggle();
    }

    public MainInputMethodService.State getCtrlState() {
        return mainInputMethodService.getCtrlState();
    }

    private int keySound(int keyCode) {
        return switch (keyCode) {
            case KeyEvent.KEYCODE_ENTER -> AudioManager.FX_KEYPRESS_RETURN;
            case KeyEvent.KEYCODE_DEL, KeyEvent.KEYCODE_FORWARD_DEL -> AudioManager.FX_KEYPRESS_DELETE;
            case KeyEvent.KEYCODE_SPACE -> AudioManager.FX_KEYPRESS_SPACEBAR;
            default -> AudioManager.FX_KEYPRESS_STANDARD;
        };
    }

    private void performInputAcceptedFeedback(int keySound) {
        AppPrefs.InputFeedback prefs = appPreferenceModel().java().getInputFeedback();
        if (prefs.getHapticEnabled().get()) {
            view.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP,
                    HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING);
        }
        if (prefs.getSoundEnabled().get()) {
            audioManager.playSoundEffect(keySound);
        }
    }

    private boolean handleKeyEventKeyCodes(int primaryCode, int keyFlags) {
        if (keyCodeIsValid(primaryCode)) {
            switch (primaryCode) {
                case KeyEvent.KEYCODE_CUT -> mainInputMethodService.cut();
                case KeyEvent.KEYCODE_COPY -> mainInputMethodService.copy();
                case KeyEvent.KEYCODE_PASTE -> mainInputMethodService.paste();
                case KeyEvent.KEYCODE_ENTER -> mainInputMethodService.commitImeOptionsBasedEnter();
                case KeyEvent.KEYCODE_DEL -> mainInputMethodService.delete();
                default -> {
                    mainInputMethodService.sendKey(primaryCode, keyFlags);
                    mainInputMethodService.resetShiftState();
                }
            }
            return true;
        }

        return false;
    }

    public void onText(CharSequence text) {
        mainInputMethodService.sendText(text.toString());
        mainInputMethodService.resetShiftState();
        performInputAcceptedFeedback(AudioManager.FX_KEYPRESS_STANDARD);
    }

    public void handleInputText(KeyboardAction keyboardAction) {
        boolean isUpperCase = mainInputMethodService.getShiftstate() != MainInputMethodService.State.OFF;
        String text = (isUpperCase && !keyboardAction.getCapsLockText().isEmpty()) ? keyboardAction.getCapsLockText() :
                keyboardAction.getText();
        onText(text);
    }

    public boolean areCharactersCapitalized() {
        return mainInputMethodService.areCharactersCapitalized();
    }

    public boolean isShiftSet() {
        return mainInputMethodService.getShiftstate() == MainInputMethodService.State.ON;
    }

    public boolean isCapsLockSet() {
        return mainInputMethodService.getShiftstate() == MainInputMethodService.State.ENGAGED;
    }

    public LayerLevel findLayer() {
        return LayerLevel.FIRST;
    }
}

File: 8vim/src/main/kotlin/inc/flide/vim8/ime/actionlisteners/KeypadActionListener.kt
package inc.flide.vim8.ime.actionlisteners

import android.content.Context
import android.media.AudioManager
import android.view.HapticFeedbackConstants
import android.view.KeyEvent
import android.view.View
import inc.flide.vim8.Vim8ImeService
import inc.flide.vim8.appPreferenceModel
import inc.flide.vim8.ime.layout.models.CustomKeycode.Companion.KEY_CODE_TO_STRING_CODE_MAP
import inc.flide.vim8.ime.layout.models.KeyboardAction
import inc.flide.vim8.ime.layout.models.LayerLevel
import inc.flide.vim8.lib.android.AndroidVersion.ATLEAST_API29_Q

@Suppress("DEPRECATION")
abstract class KeypadActionListener(
    protected val vim8ImeService: Vim8ImeService,
    protected val view: View
) {
    private val prefs by appPreferenceModel()
    val ctrlState: Boolean
        get() = vim8ImeService.ctrlState
    val isPassword: Boolean
        get() = vim8ImeService.isPassword
    private val audioManager: AudioManager =
        view.context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

    private fun keyCodeIsValid(keyCode: Int): Boolean {
        val keycodeProfileSwitch =
            if (ATLEAST_API29_Q) KeyEvent.KEYCODE_PROFILE_SWITCH else KEYCODE_PROFILE_SWITCH
        return keyCode >= KeyEvent.KEYCODE_UNKNOWN && keyCode <= keycodeProfileSwitch
    }

    fun performCtrlToggle() {
        vim8ImeService.performCtrlToggle()
    }
    fun handleInputKey(keyCode: Int, keyFlags: Int) {
        val actionHandled =
            handleKeyEventKeyCodes(keyCode, keyFlags) || (
                KEY_CODE_TO_STRING_CODE_MAP[keyCode]?.handleKeyCode(
                    vim8ImeService
                ) ?: true
                )
        if (!actionHandled) {
            onText(keyCode.toChar().toString())
        } else {
            performInputAcceptedFeedback(keySound(keyCode))
        }
    }

    private fun keySound(keyCode: Int): Int {
        return when (keyCode) {
            KeyEvent.KEYCODE_ENTER -> AudioManager.FX_KEYPRESS_RETURN
            KeyEvent.KEYCODE_DEL, KeyEvent.KEYCODE_FORWARD_DEL -> AudioManager.FX_KEYPRESS_DELETE
            KeyEvent.KEYCODE_SPACE -> AudioManager.FX_KEYPRESS_SPACEBAR
            else -> AudioManager.FX_KEYPRESS_STANDARD
        }
    }

    private fun performInputAcceptedFeedback(keySound: Int) {
        if (prefs.inputFeedback.hapticEnabled.get()) {
            view.performHapticFeedback(
                HapticFeedbackConstants.KEYBOARD_TAP,
                HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING
            )
        }
        if (prefs.inputFeedback.soundEnabled.get()) {
            audioManager.playSoundEffect(keySound)
        }
    }

    private fun handleKeyEventKeyCodes(primaryCode: Int, keyFlags: Int): Boolean {
        if (keyCodeIsValid(primaryCode)) {
            when (primaryCode) {
                KeyEvent.KEYCODE_CUT -> vim8ImeService.cut()
                KeyEvent.KEYCODE_COPY -> vim8ImeService.copy()
                KeyEvent.KEYCODE_PASTE -> vim8ImeService.paste()
                KeyEvent.KEYCODE_ENTER -> vim8ImeService.commitImeOptionsBasedEnter()
//                KeyEvent.KEYCODE_DEL -> vim8ImeService.delete()
                else -> {
                    val flags = if (isDPad(primaryCode)) {
                        keyFlags or vim8ImeService.ctrlFlag
                    } else {
                        keyFlags
                    }
                    vim8ImeService.sendKey(primaryCode, flags)
                    vim8ImeService.resetShiftState()
                }
            }
            return true
        }
        return false
    }

    private fun isDPad(primaryCode: Int): Boolean {
        return when (primaryCode) {
            KeyEvent.KEYCODE_DPAD_UP,
            KeyEvent.KEYCODE_DPAD_DOWN,
            KeyEvent.KEYCODE_DPAD_LEFT,
            KeyEvent.KEYCODE_DPAD_RIGHT -> {
                true
            }

            else -> false
        }
    }

    fun onText(text: CharSequence) {
        vim8ImeService.sendText(text.toString())
        vim8ImeService.resetShiftState()
        performInputAcceptedFeedback(AudioManager.FX_KEYPRESS_STANDARD)
    }

    fun handleInputText(keyboardAction: KeyboardAction) {
        val isUpperCase = vim8ImeService.shiftState != Vim8ImeService.State.OFF
        val text =
            if (isUpperCase && keyboardAction.capsLockText.isNotEmpty()) {
                keyboardAction.capsLockText
            } else {
                keyboardAction.text
            }
        onText(text)
    }

    fun areCharactersCapitalized(): Boolean {
        return vim8ImeService.areCharactersCapitalized()
    }

    val isShiftSet: Boolean
        get() = vim8ImeService.shiftLockFlag == KeyEvent.META_SHIFT_ON
    val isCapsLockSet: Boolean
        get() = vim8ImeService.capsLockFlag == KeyEvent.META_CAPS_LOCK_ON

    open fun findLayer(): LayerLevel {
        return LayerLevel.FIRST
    }

    companion object {
        const val KEYCODE_PROFILE_SWITCH = 288
    }
}
